### PR TITLE
Fix single file uploads in multi-file forms

### PR DESCRIPTION
**Pull request by pydanny**
---
name: Pull Request
about: Propose changes to the codebase
title: '[PR] Fix single file uploads in multi-file forms'
labels: 'bug'
assignees: ''

---

**Related Issue**

#513 Uploading a single file on a multiple file field requires a try/except block

**Proposed Changes**

If the annotation for the response element says it is a `GenericAlias` (which roughly means 'iterable') but the response element isn't actually iterable, then wrap it in a list.

**Types of changes**
What types of changes does your code introduce? Put an `x` in all the boxes that apply:
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

**Checklist**
Go over all the following points, and put an `x` in all the boxes that apply:
- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.
- [x] I am aware that this is an nbdev project, and I have edited, cleaned, and synced the source notebooks instead of editing .py or .md files directly.

**Additional Information**
Any additional information, configuration or data that might be necessary to reproduce the issue.

### CODE REVIEW

**Comment by pydanny** [comment]
Hey @jph00, here's a patch for multi-file uploads that comes with a test.

File: fasthtml/core.py [hunk 1]
```diff
@@ -191,6 +191,10 @@ async def _find_p(req, arg:str, p:Parameter):
     if (res in (empty,None)) and p.default is empty: raise HTTPException(400, f"Missing required field: {arg}")
     # If we have a default, return that if we have no value
     if res in (empty,None): res = p.default
+    # if anno is a Generic Alias but res isn't an iterable then wrap in list
+    if isinstance(anno, GenericAlias):
```
Comment (jph00) [inline fasthtml/core.py:195]: I think there are parameterised generics that aren't lists? We should probably me more specific about what we're looking for.

**Review by jph00 — CHANGES_REQUESTED** [review]
Thanks Danny! Just one suggestion to look at.

File: fasthtml/core.py (hunk 1 above)
Comment (pydanny) [reply fasthtml/core.py]: @jph00 The challenge is that for this error `anno` is a `GenericAlias`, which arguably is a bug in its own right. Or maybe we're just not particular enough so this is a missing feature? Looking into that now.

In the meantime, I've changed the `try/except` to a more precise check. Hopefully that goes away, replaced with a more accurate `anno` specification.

Comment (pydanny) [reply fasthtml/core.py]: Diving into Fastcore's [signature_ex](https://fastcore.fast.ai/basics.html#signature_ex) function.

Comment (pydanny) [reply fasthtml/core.py]: Hmmm... I think `signature_ex` is a tangent. Digging into the behavior around params in `_wrap_req`.

**Pull request closed without merging**